### PR TITLE
Enable quidem tests to use different suppliers

### PR DESCRIPTION
**Pull request by kgyrtkirk**
* enable quidem uri support for `druidtest:///?ComponentSupplier=Nested` and similar
* changes the way `SqlTestFrameworkConfig` is being applied; all options will have their own annotation (its kinda impossible to detect that an annotation has a set value or its the default)
* enables hierarchical processing of config annotation (was needed to enable class level supplier annotation)
* moves uri processing related string2config stuff into `SqlTestFrameworkConfig`

### CODE REVIEW

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java:102]: I've renamed it.
I believe the reason I've shortened this to `Supplier` was due to the fact that these could be specified in the jdbc uri; and long key names could take up a lot of extra line-space - at the cost of readability.

Right now I've "accepted" to follow a convention of only accepting the configuration key if it fully matches with the annotation's name; so`druidtest:///?ComponentSupplier=Something` instead of `druidtest:///?componentsupplier=Something` or  `druidtest:///?Componentsupplier=Something`  ; I was wondering about adding the ability to match the options in an ignore case manner...but writing all this down....it might be more annoying to typo an option name (whenever its short or long) and not getting the expected behaviour....maybe I should instead add a validation for the provided config keys

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java:166]: yes; only classes under the `org.apache.druid` will be considered - what do you suggest? 
* use `ServiceLoader` to register implementations - I would prefer this ; but seems unnecessarily complication
* scan all packages?
* something else?

I've replaced it with the following logic:
* introduced a `LoadingCache` which is keyed by `packageName`
* lookup first tries the `org.apache.druid.sql.calcite` package
* falls back to asks for the root package - and if there are no matches even there -> it returns with an error

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java:301]: The `AvaticaServer` related things are added by wrapping the supplier with a delegate; in normal operation that's not needed.
I think that could be removed later - as it would be better to package that part into a module or something

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java:150]: unfortunately - some of these options have no close relation to the `QueryContext`

Other way around could be to move these processing things next to the options - but that seemed a bit too much at first

Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java]: I've introduced a set of classes/processor/etc to disperse this method to the config options...its different for sure :)

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (rohangarg) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java]: Thanks for the clarification and changes! And yes, I think adding a validation for config keys in future might be good - since these configs are only used in tests, hence I think it makes sense for us to be strict in accepting them.

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (rohangarg) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java:96]: Ok, I see... the current setup makes sense if the configs allowed are only for the environment. 
And I agree that removing the `minTopNThreashold` from this syntax would be helpful in future - to avoid bad culture.
Although, I'm not sure if I understand how someone can specify the context properties right now for a test. Is it possible as of now to do that?

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (rohangarg) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java]: Thanks for the new set of changes, they look good to me :+1:

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (rohangarg) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java]: The current implementation of using a chained method with calcite and root package, along with the caching looks good to me 👍  One more thing to look at for performance changes could be to check the runtimes of SQL module tests in master vs this PR - hopefully they should be pretty close.

**Review by rohangarg — APPROVED** [review]
Thanks for the updates @kgyrtkirk 
LGTM :+1: 

Approving the changes, but waiting for your response on a couple of comments before merging

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java:96]: by using `testBuilder().queryContext()` additional querycontext variables can be set (see an example [here](https://github.com/apache/druid/blob/dded473ac0a87e67033cd5dbbe67f1c9fa4b335d/sql/src/test/java/org/apache/druid/sql/calcite/CalciteWindowQueryTest.java#L202-L206))


looking closer at `TopNQueryConfig` ; there seems to be 2 things:
* **(A)** there is a `TopNQueryConfig` [class](https://github.com/apache/druid/blob/dded473ac0a87e67033cd5dbbe67f1c9fa4b335d/processing/src/main/java/org/apache/druid/query/topn/TopNQueryConfig.java#L34) which is jackon serializable
* **(B)** but `TopNQueryToolChest` also enables a `QueryContext` key to override that value [here](https://github.com/apache/druid/blob/dded473ac0a87e67033cd5dbbe67f1c9fa4b335d/processing/src/main/java/org/apache/druid/query/topn/TopNQueryQueryToolChest.java#L629-L630)

I think the test system was probably built to configure the `A` approach and later probably `B` was also added...
since querycontext defaults can be configured globally - I think keeping the `A` might be redundant ; and it might worth considering the removal of it.

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java]: this newly added by-name lookup will not interfere with existing tests; as in those cases a ComponentSupplier class is  specified as a java class explicitly in the annotation - however in case it comes from a jdbc-uri it will need the lookup.

run all `Calcite*Test` on master and on this branch - didn't seen any relevant difference in the runtimes
<details>
<summary>testrun on master</summary>

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.druid.sql.calcite.CalciteNestedDataQueryTest
[INFO] Tests run: 145, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.255 s -- in org.apache.druid.sql.calcite.CalciteNestedDataQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteArraysQueryTest
[INFO] Tests run: 156, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 11.42 s -- in org.apache.druid.sql.calcite.CalciteArraysQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteUnionQueryTest
[INFO] Tests run: 14, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.881 s -- in org.apache.druid.sql.calcite.CalciteUnionQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteSysQueryTest
[WARN] Tests run: 4, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 2.646 s -- in org.apache.druid.sql.calcite.CalciteSysQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteLookupFunctionQueryTest
[INFO] Tests run: 90, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.361 s -- in org.apache.druid.sql.calcite.CalciteLookupFunctionQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteStrictInsertTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.488 s -- in org.apache.druid.sql.calcite.CalciteStrictInsertTest
[INFO] Running org.apache.druid.sql.calcite.CalciteSimpleQueryTest
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.734 s -- in org.apache.druid.sql.calcite.CalciteSimpleQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteParameterQueryTest
[INFO] Tests run: 20, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.874 s -- in org.apache.druid.sql.calcite.CalciteParameterQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteJoinQueryTest
[WARN] Tests run: 586, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 16.89 s -- in org.apache.druid.sql.calcite.CalciteJoinQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteCatalogInsertTest
[INFO] Tests run: 19, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.637 s -- in org.apache.druid.sql.calcite.CalciteCatalogInsertTest
[INFO] Running org.apache.druid.sql.calcite.CalciteExplainQueryTest
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.514 s -- in org.apache.druid.sql.calcite.CalciteExplainQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteExportTest
[WARN] Tests run: 13, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 2.472 s -- in org.apache.druid.sql.calcite.CalciteExportTest
[INFO] Running org.apache.druid.sql.calcite.CalciteWindowQueryTest
[WARN] Tests run: 41, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 4.641 s -- in org.apache.druid.sql.calcite.CalciteWindowQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteTableAppendTest
[INFO] Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.469 s -- in org.apache.druid.sql.calcite.CalciteTableAppendTest
[INFO] Running org.apache.druid.sql.calcite.CalciteReplaceDmlTest
[INFO] Tests run: 46, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.769 s -- in org.apache.druid.sql.calcite.CalciteReplaceDmlTest
[INFO] Running org.apache.druid.sql.calcite.planner.CalcitePlannerModuleTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.435 s -- in org.apache.druid.sql.calcite.planner.CalcitePlannerModuleTest
[INFO] Running org.apache.druid.sql.calcite.planner.CalcitesTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 s -- in org.apache.druid.sql.calcite.planner.CalcitesTest
[INFO] Running org.apache.druid.sql.calcite.CalciteCorrelatedQueryTest
[INFO] Tests run: 35, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.919 s -- in org.apache.druid.sql.calcite.CalciteCorrelatedQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteQueryTest
[WARN] Tests run: 389, Failures: 0, Errors: 0, Skipped: 4, Time elapsed: 13.97 s -- in org.apache.druid.sql.calcite.CalciteQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteSubqueryTest
[INFO] Tests run: 44, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.449 s -- in org.apache.druid.sql.calcite.CalciteSubqueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteInsertDmlTest
[INFO] Tests run: 59, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.959 s -- in org.apache.druid.sql.calcite.CalciteInsertDmlTest
[INFO] Running org.apache.druid.sql.calcite.CalciteSelectQueryTest
[INFO] Tests run: 59, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.949 s -- in org.apache.druid.sql.calcite.CalciteSelectQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteMultiValueStringQueryTest
[INFO] Tests run: 59, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.020 s -- in org.apache.druid.sql.calcite.CalciteMultiValueStringQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteCatalogReplaceTest
[INFO] Tests run: 19, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.566 s -- in org.apache.druid.sql.calcite.CalciteCatalogReplaceTest
[INFO] Running org.apache.druid.sql.calcite.CalciteTimeBoundaryQueryTest
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.472 s -- in org.apache.druid.sql.calcite.CalciteTimeBoundaryQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteScanSignatureTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.472 s -- in org.apache.druid.sql.calcite.CalciteScanSignatureTest
[INFO] 
[INFO] Results:
[INFO] 
[WARN] Tests run: 1845, Failures: 0, Errors: 0, Skipped: 11
[INFO] 
[INFO] 
[INFO] --- animal-sniffer-maven-plugin:1.23:check (check-java-api) @ druid-sql ---
[INFO] Checking unresolved references to org.codehaus.mojo.signature:java18:1.0
[INFO] Segment walltime 124 s, segment projects service time 124 s, effective/maximum degree of concurrency 1.00/11
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:04 min (Wall Clock)
[INFO] Finished at: 2024-05-08T10:54:19Z
[INFO] ------------------------------------------------------------------------
```

</details>

<details>
<summary>testrun on branch</summary>

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.druid.sql.calcite.CalciteNestedDataQueryTest
[INFO] Tests run: 145, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.195 s -- in org.apache.druid.sql.calcite.CalciteNestedDataQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteArraysQueryTest
[INFO] Tests run: 156, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 11.17 s -- in org.apache.druid.sql.calcite.CalciteArraysQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteUnionQueryTest
[INFO] Tests run: 14, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.815 s -- in org.apache.druid.sql.calcite.CalciteUnionQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteSysQueryTest
[WARN] Tests run: 4, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 2.628 s -- in org.apache.druid.sql.calcite.CalciteSysQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteLookupFunctionQueryTest
[INFO] Tests run: 90, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.194 s -- in org.apache.druid.sql.calcite.CalciteLookupFunctionQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteStrictInsertTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.519 s -- in org.apache.druid.sql.calcite.CalciteStrictInsertTest
[INFO] Running org.apache.druid.sql.calcite.CalciteSimpleQueryTest
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.774 s -- in org.apache.druid.sql.calcite.CalciteSimpleQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteParameterQueryTest
[INFO] Tests run: 20, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.923 s -- in org.apache.druid.sql.calcite.CalciteParameterQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteJoinQueryTest
[WARN] Tests run: 586, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 16.70 s -- in org.apache.druid.sql.calcite.CalciteJoinQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteCatalogInsertTest
[INFO] Tests run: 19, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.681 s -- in org.apache.druid.sql.calcite.CalciteCatalogInsertTest
[INFO] Running org.apache.druid.sql.calcite.CalciteExplainQueryTest
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.536 s -- in org.apache.druid.sql.calcite.CalciteExplainQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteExportTest
[WARN] Tests run: 13, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 2.483 s -- in org.apache.druid.sql.calcite.CalciteExportTest
[INFO] Running org.apache.druid.sql.calcite.CalciteWindowQueryTest
[WARN] Tests run: 41, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 4.509 s -- in org.apache.druid.sql.calcite.CalciteWindowQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteTableAppendTest
[INFO] Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.490 s -- in org.apache.druid.sql.calcite.CalciteTableAppendTest
[INFO] Running org.apache.druid.sql.calcite.CalciteReplaceDmlTest
[INFO] Tests run: 46, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.813 s -- in org.apache.druid.sql.calcite.CalciteReplaceDmlTest
[INFO] Running org.apache.druid.sql.calcite.planner.CalcitePlannerModuleTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.428 s -- in org.apache.druid.sql.calcite.planner.CalcitePlannerModuleTest
[INFO] Running org.apache.druid.sql.calcite.planner.CalcitesTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s -- in org.apache.druid.sql.calcite.planner.CalcitesTest
[INFO] Running org.apache.druid.sql.calcite.CalciteCorrelatedQueryTest
[INFO] Tests run: 35, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.061 s -- in org.apache.druid.sql.calcite.CalciteCorrelatedQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteQueryTest
[WARN] Tests run: 389, Failures: 0, Errors: 0, Skipped: 4, Time elapsed: 14.12 s -- in org.apache.druid.sql.calcite.CalciteQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteSubqueryTest
[INFO] Tests run: 44, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.475 s -- in org.apache.druid.sql.calcite.CalciteSubqueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteInsertDmlTest
[INFO] Tests run: 59, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.018 s -- in org.apache.druid.sql.calcite.CalciteInsertDmlTest
[INFO] Running org.apache.druid.sql.calcite.CalciteSelectQueryTest
[INFO] Tests run: 59, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.973 s -- in org.apache.druid.sql.calcite.CalciteSelectQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteMultiValueStringQueryTest
[INFO] Tests run: 59, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.062 s -- in org.apache.druid.sql.calcite.CalciteMultiValueStringQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteCatalogReplaceTest
[INFO] Tests run: 19, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.681 s -- in org.apache.druid.sql.calcite.CalciteCatalogReplaceTest
[INFO] Running org.apache.druid.sql.calcite.CalciteTimeBoundaryQueryTest
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.577 s -- in org.apache.druid.sql.calcite.CalciteTimeBoundaryQueryTest
[INFO] Running org.apache.druid.sql.calcite.CalciteScanSignatureTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.448 s -- in org.apache.druid.sql.calcite.CalciteScanSignatureTest
[INFO] 
[INFO] Results:
[INFO] 
[WARN] Tests run: 1845, Failures: 0, Errors: 0, Skipped: 11
[INFO] 
[INFO] 
[INFO] --- animal-sniffer-maven-plugin:1.23:check (check-java-api) @ druid-sql ---
[INFO] Checking unresolved references to org.codehaus.mojo.signature:java18:1.0
[INFO] Segment walltime 126 s, segment projects service time 126 s, effective/maximum degree of concurrency 1.00/11
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:06 min (Wall Clock)
[INFO] Finished at: 2024-05-08T10:52:09Z
[INFO] ------------------------------------------------------------------------
```
</details>

Comment (rohangarg) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java]: ah, I see. Sorry I missed that! Although, the few quidem tests we have right now would invoke the lookup right?
Also, thanks for checking the time difference between the runtimes

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (rohangarg) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java:96]: Thanks for the pointer for queryContext :+1: 

Oh I see, yes your explanation for `TopNQueryConfig` does make sense to me. And agree that we could remove it in the future in exchange for the context param - but we'd need to find a way to support the old config name too without the explicit config class for it.

File: sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java
Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/SqlTestFrameworkConfig.java]: yes - they use this method; and they will benefit from these caches more-and-more as we will get more cases!

**Pull request merged**
